### PR TITLE
Get almost all the tests working

### PR DESCRIPTION
**Pull request by nikihowe**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

- pass all the tests (except for one of the amazons ones)

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [ ] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: pypolygames/tests/test_zoo.py [hunk 1]
```diff
@@ -6,11 +6,42 @@
 from unittest import SkipTest
 import pytest
 import torch
+import torch.nn.functional as F
 from .. import model_zoo
 from ..model_zoo.utils import get_game_info
 from .. import params
 from .. import utils
 
+# NOTE: Copied this loss from a model in the first commit
```
Comment (rhaps0dy) [inline pypolygames/tests/test_zoo.py]: Bit of a strange test, but it will do . It has a bunch of asserts indeed.

Without a loss how are the NNs trained with this code base? Maybe we don't care because we're just going to have to write an adversarial training part anyways.

File: pypolygames/tests/test_zoo.py [hunk 2]
```diff
@@ -30,10 +61,11 @@ def test_models(model_name) -> None:
...
     # loss
     multi_counter = utils.MultiCounter(root=None)
-    pi_mask = torch.ones(outputs["pi"].shape)
-    model.loss(
-        model, input_data, outputs["v"], outputs["pi"], pi_mask, multi_counter
-    )  # make sure it computes something
+    pi_mask = torch.ones(outputs["pi_logit"].shape)
+    
+    loss(
+        model, input_data, outputs["v"], outputs["pi_logit"], pi_mask, multi_counter
```
Comment (rhaps0dy) [inline pypolygames/tests/test_zoo.py:71]: you're passing a logit to the loss above but it takes a policy `pi`.

File: src/games/hex.h [hunk 3]
```diff
@@ -17,7 +17,8 @@
 
 namespace Hex {
 
-enum Color { COLOR_BLACK, COLOR_WHITE, COLOR_NONE };
+// Niki added the explicit 0
```
Comment (rhaps0dy) [inline src/games/hex.h:20]: I don't think you need to add comments stating what you changed, but seems fine.

File: src/games/hex_state.h [hunk 4]
```diff
@@ -80,14 +80,16 @@ void Hex::State<SIZE, PIE>::ApplyAction(const _Action& action) {
...
+  assert(j >= 0 && j < SIZE);
   int index = _board.convertCellToIndex(Cell(i, j));
   std::optional<int> lastIndex = _board.getLastIndex();
 
   // update features
   // TODO assert action is in legal actions ?
   if (not lastIndex or *lastIndex != index) {
     Color currentColor = _board.getCurrentColor();
-    _features.at(((currentColor * SIZE) + i) * SIZE + j) = 1.f;
+    _features.at((((currentColor) * SIZE) + i) * SIZE + j) = 1.f;
```
Comment (rhaps0dy) [inline src/games/hex_state.h]: strange extra parens?
```suggestion
    _features.at(((currentColor * SIZE) + i) * SIZE + j) = 1.f;
```

File: src/mcts/old_test.cc
Comment (rhaps0dy) [inline src/mcts/old_test.cc:1]: I recommend deleting this to prevent further confusion. Old files are always available through the commit history.

File: tests/connectfour-tests.cc [hunk 5]
```diff
@@ -13,6 +13,48 @@
...
+    // Fill the first column with alternating pieces
+    for (int i = 0; i < 6; ++i) {
+        _Action action(0, 0, 0, 0);  // play in column 0
+        state.ApplyAction(action);
+        // std::cout << "played action 0, now board is" << std::endl;
+        // state.printCurrentBoard();
+        // printAllowedActions(state);
+    }
+
+    // At this point, the board should look as follows
```
Comment (rhaps0dy) [inline tests/connectfour-tests.cc:45]: seems like a prime candidate for an assert :)

File: tests/havannah-state-tests.cc [hunk 6]
```diff
@@ -52,6 +52,7 @@ TEST(HavannahStateGroup, init_0) {
  const int nbActions = fullsize*fullsize - size*(size-1);  // 9 * 9 - 5 * 4 = 81 - 20 = 61
 
  Havannah::StateTest<size, true, false> state(0, history, turnFeatures);
+ state.Initialize();
```
Comment (rhaps0dy) [inline tests/havannah-state-tests.cc:55]: Why not run this in the constructor to `StateTest` ?

File: tests/havannah-state-tests.cc [hunk 7]
```diff
@@ -208,7 +218,7 @@ TEST(HavannahStateGroup, init_2) {
   ASSERT_EQ(0, action.GetX());
   ASSERT_EQ(i, action.GetY());
   ASSERT_EQ(j, action.GetZ());
-  ASSERT_EQ(h, action.GetHash());
+  ASSERT_EQ(0, action.GetHash()); // Niki changed to 0
```
Comment (rhaps0dy) [inline tests/havannah-state-tests.cc]: Is the hash always 0? What is this testing? Maybe you should just remove `getHash` entirely.

File: tests/havannah-state-tests.cc (hunk 6 above)
Comment (rhaps0dy) [reply tests/havannah-state-tests.cc:55]: Maybe `setFeatures(&opt)` should also be in the constructor.

File: tests/hex-state-tests.cc [hunk 8]
```diff
@@ -7,445 +7,421 @@
...
-  auto a = &state.GetLegalActions()[k];
-  ASSERT_EQ(0, a->GetX());
-  ASSERT_EQ(i, a->GetY());
-  ASSERT_EQ(j, a->GetZ());
-  ASSERT_EQ(k, a->GetHash());
-  ASSERT_EQ(k, a->GetIndex());
- }
-}
+  Hex::StateTest<7, true> state(0, 0, false);
+  state.Initialize();  // Niki added
```
Comment (rhaps0dy) [inline tests/hex-state-tests.cc:40]: This one should also probably go in the constructor

**Review by rhaps0dy — CHANGES_REQUESTED** [review]

File: pypolygames/tests/test_zoo.py (hunk 1 above)
Comment (nikihowe) [reply pypolygames/tests/test_zoo.py]: There is a loss ([here](https://github.com/AlignmentResearch/polygames/blob/main/pypolygames/model_zoo/loss.py)) just it doesn't belong to the model, but rather to the ModelManager I believe.

File: pypolygames/tests/test_zoo.py (hunk 2 above)
Comment (nikihowe) [reply pypolygames/tests/test_zoo.py:71]: I think the "pi" in loss.py is also an array. So effectively it could be called "pi_logits". I left it as "pi" since that's how it was originally. Do you think it would be better to change it to "pi_logits"?

File: tests/connectfour-tests.cc (hunk 5 above)
Comment (nikihowe) [reply tests/connectfour-tests.cc:45]: Working on it

File: tests/havannah-state-tests.cc (hunk 7 above)
Comment (nikihowe) [reply tests/havannah-state-tests.cc]: Do you think this should be done before merging this? Or just in general?

File: src/mcts/old_test.cc
Comment (nikihowe) [reply src/mcts/old_test.cc:1]: So we accept abandoning TicTacToe?

File: tests/connectfour-tests.cc (hunk 5 above)
Comment (nikihowe) [reply tests/connectfour-tests.cc:45]: Done

File: tests/havannah-state-tests.cc (hunk 6 above)
Comment (nikihowe) [reply tests/havannah-state-tests.cc:55]: I think you have done this?

File: tests/hex-state-tests.cc (hunk 8 above)
Comment (nikihowe) [reply tests/hex-state-tests.cc:40]: I think you have done this too?

File: pypolygames/__init__.py [hunk 9]
```diff
@@ -25,3 +25,4 @@
 mcts = os.path.join(root, "build/src", "mcts")
 if mcts not in sys.path:
     sys.path.append(mcts)
+
```
Comment (tomtseng) [inline pypolygames/__init__.py]: nit: remove changes to this file

File: pypolygames/tests/test_mcts.py [hunk 10]
```diff
@@ -37,6 +37,8 @@ def test_mcts(game_name) -> None:
         "ChineseCheckers",
     ]
     too_bad = ["Havannah5", "Havannah5pie", "Surakarta", "DiceShogi", "Connect6"]
+    if is_one_player_game:
+        raise SkipTest("One player game -- we're not interested")
     if game_name in crashing + too_slow:  # + one_player_games:
```
Comment (tomtseng) [inline pypolygames/tests/test_mcts.py]:
```suggestion
        raise SkipTest(f"Skipping {game_name} (one-player game)")
    if game_name in crashing + too_slow:
```

File: pypolygames/tests/test_zoo.py (hunk 1 above)
Comment (tomtseng) [inline pypolygames/tests/test_zoo.py]: not sure what "first commit" refers to. Maybe comment should paste a GitHub link to the commit & line of code? e.g., link to https://github.com/facebookarchive/Polygames/blob/eb5390e57cc38e5287bf6dcfb420308a5995d194/pypolygames/model_zoo/amazons_model.py#L135 or whatever it is you copied this from

File: src/games/hex_state.h [hunk 11]
```diff
@@ -80,6 +80,8 @@ void Hex::State<SIZE, PIE>::ApplyAction(const _Action& action) {
   // find board move from action
   int i = action.GetY();
   int j = action.GetZ();
+  assert(i >= 0 && i < SIZE);  // N added
```
Comment (tomtseng) [inline src/games/hex_state.h:83]: eh no need to comment `// N added`, we can see you added it in the Git blame

File: src/mcts/old_test.cc [hunk 12]
```diff
@@ -0,0 +1,290 @@
+/**
```
Comment (tomtseng) [inline src/mcts/old_test.cc]: is this file unused? i'd just delete this file if you're not using it

**Review by tomtseng — COMMENTED** [review]
(didn't look at the C++ tests yet)

Is the name of this PR accurate? It's making me a bit confused about what this PR is aiming to do

File: pypolygames/__init__.py (hunk 9 above)
Comment (nikihowe) [reply pypolygames/__init__.py]: hmm, isn't it better to have a newline at the end of files?

**Comment by nikihowe** [comment]
> (didn't look at the C++ tests yet)
> 
> Is the name of this PR accurate? It's making me a bit confused about what this PR is aiming to do

Nice point. I updated it.

File: pypolygames/tests/test_zoo.py (hunk 1 above)
Comment (nikihowe) [reply pypolygames/tests/test_zoo.py]: Done, thank you

File: src/games/hex_state.h (hunk 11 above)
Comment (nikihowe) [reply src/games/hex_state.h:83]: Should I remove all mentions of things that I did, and leave them all to the Git blame? It seemed useful to me to see at a glance why things are a certain way.

File: src/mcts/old_test.cc (hunk 12 above)
Comment (nikihowe) [reply src/mcts/old_test.cc]: Well, we're not using it now, but it would be nice to get the tictactoe tests working (which is what these are). Originally this was the test file but I swapped out tictactoe for connectfour.

File: pypolygames/tests/test_zoo.py (hunk 2 above)
Comment (rhaps0dy) [reply pypolygames/tests/test_zoo.py:71]: Softmax it before passing in or something?

File: tests/havannah-state-tests.cc (hunk 7 above)
Comment (rhaps0dy) [reply tests/havannah-state-tests.cc]: Yeah before merging this, as we already changed tests for it in this PR

File: src/mcts/old_test.cc
Comment (rhaps0dy) [reply src/mcts/old_test.cc:1]: Yea nuke it!!

File: tests/havannah-state-tests.cc (hunk 7 above)
Comment (rhaps0dy) [inline tests/havannah-state-tests.cc]:
```suggestion
  ASSERT_EQ(0, action.GetHash()); // TODO remove GetHash completely
```

File: pypolygames/__init__.py (hunk 9 above)
Comment (tomtseng) [reply pypolygames/__init__.py]: The last line of a file should end with a newline (like all other lines in the file). But in this case we're adding an additional line to the file that's just a newline, which is unnecessary

File: src/games/hex_state.h (hunk 11 above)
Comment (tomtseng) [reply src/games/hex_state.h:83]: my preference would be to remove the comments and just use Git blame because I think it leaves the code cleaner. But if you find leaving the comments to be useful then I think having them is NBD, it's understandable to me

File: src/mcts/old_test.cc (hunk 12 above)
Comment (tomtseng) [reply src/mcts/old_test.cc]: I see, in that case I'd leave a TODO comment at the top of the file saying that this file isn't in use but we should get it working in the future

**Review by tomtseng — APPROVED** [review]

File: pypolygames/tests/test_zoo.py (hunk 2 above)
Comment (nikihowe) [reply pypolygames/tests/test_zoo.py:71]: I'll fix this in a later PR

File: tests/havannah-state-tests.cc (hunk 7 above)
Comment (nikihowe) [reply tests/havannah-state-tests.cc]: Ok, I have done this

File: pypolygames/__init__.py (hunk 9 above)
Comment (nikihowe) [reply pypolygames/__init__.py]: I think I must be missing something. It looks to me like the last line of code is 27, and that we are adding a newline on 28

File: src/mcts/old_test.cc (hunk 12 above)
Comment (nikihowe) [reply src/mcts/old_test.cc]: I simply raised an issue

File: pypolygames/__init__.py (hunk 9 above)
Comment (rhaps0dy) [reply pypolygames/__init__.py]: Yes, there should be a newline at the end of 27, but nothing in line 28. Here there are two newlines at the end of "27".

Comment (rhaps0dy) [inline pypolygames/__init__.py]:
```suggestion
```

Comment (rhaps0dy) [reply pypolygames/__init__.py]: This is what `black` would do.

Code style reviews are a waste of time and we should just use a formatter like `black`.

Comment (nikihowe) [reply pypolygames/__init__.py]: I see. I was (don't remember why) under the impression that there should be an empty line (so, equivalent to two newlines) and the bottom of a file.

File: tests/havannah-state-tests.cc (hunk 7 above)
Comment (nikihowe) [reply tests/havannah-state-tests.cc]: I simply removed the hash completely

File: pypolygames/__init__.py (hunk 9 above)
Comment (nikihowe) [reply pypolygames/__init__.py]: Nice idea, how do we set it up?

Comment (nikihowe) [reply pypolygames/__init__.py]: Also I fixed the newline thing, I think we're saying the same thing

Comment (tomtseng) [reply pypolygames/__init__.py]: if you `pip install black`, you run `black <filename>` to lint a file

If you want to add it as a GitHub check too then you do it through CircleCI (`go_attack` does this: https://github.com/AlignmentResearch/go_attack/blob/main/.circleci/config.yml#L59-L76)

**Review by rhaps0dy — APPROVED** [review]
